cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp

```diff
@@ -57,6 +57,132 @@ inline void validate_input(const sycl::queue& q,
+                  "Dimensions of covariance matrix must match feature count");
+    ONEDAL_ASSERT(sums.get_dimension(0) == data.get_dimension(1),
+                  "Element count of sums must match feature count");
+    ONEDAL_ASSERT(vars.get_dimension(0) == data.get_dimension(1),
+                  "Element count of vars must match feature count");
+    ONEDAL_ASSERT(means.get_dimension(0) == data.get_dimension(1),
+                  "Element count of means must match feature count");
+    ONEDAL_ASSERT(is_known_usm(q, sums.get_data()));
+    ONEDAL_ASSERT(is_known_usm(q, data.get_data()));
+    ONEDAL_ASSERT(is_known_usm(q, cov.get_mutable_data()));
+    ONEDAL_ASSERT(is_known_usm(q, means.get_mutable_data()));
+    ONEDAL_ASSERT(is_known_usm(q, vars.get_mutable_data()));
+}
+
+template <typename Float>
+inline sycl::event compute_means(sycl::queue& q,
+                                 const ndview<Float, 2>& data,
+                                 const ndview<Float, 1>& sums,
+                                 ndview<Float, 1>& means,
+                                 const event_vector& deps) {
+    const std::int64_t column_count = data.get_dimension(1);
+    const std::int64_t row_count = data.get_dimension(0);
+
+    const Float inv_n = Float(1.0 / double(row_count));
+
+    const Float* sums_ptr = sums.get_data();
+    Float* means_ptr = means.get_mutable_data();
+
+    return q.submit([&](sycl::handler& cgh) {
+        const auto range = ::oneapi::dal::backend::make_multiple_nd_range_1d(
+            column_count,
+            ::oneapi::dal::backend::device_max_wg_size(q));
+
+        cgh.depends_on(deps);
+        cgh.parallel_for(range, [=](sycl::nd_item<1> id) {
+            const std::int64_t i = id.get_global_id();
+            if (i < column_count) {
+                const Float s = sums_ptr[i];
+                means_ptr[i] = inv_n * s;
+            }
+        });
+    });
+}
+
+template <typename Float>
+inline sycl::event prepare_covariance(sycl::queue& q,
+                                      std::int64_t row_count,
+                                      const ndview<Float, 1>& sums,
+                                      const ndview<Float, 2>& cov,
+                                      ndview<Float, 1>& means,
+                                      ndview<Float, 1>& vars,
+                                      const event_vector& deps) {
+    const std::int64_t n = row_count;
+    const std::int64_t p = sums.get_count();
+    const Float inv_n = Float(1.0 / double(n));
+    const Float inv_n1 = (n > 1.0f) ? Float(1.0 / double(n - 1)) : 1.0f;
+
+    const Float* sums_ptr = sums.get_data();
+    const Float* cov_ptr = cov.get_mutable_data();
+    Float* means_ptr = means.get_mutable_data();
+    Float* vars_ptr = vars.get_mutable_data();
+
+    return q.submit([&](sycl::handler& cgh) {
+        const auto range = make_multiple_nd_range_1d(p, device_max_wg_size(q));
```

Comment:
Please use normal ranges (not `nd_range`s) in such kernels. It helps to avoid many issues:
1) You should care no more about wg size and etc.
2) You can write code that will not use `if (i < p)` pattern and also variable `p` will not be captured via kernel